File: packages/manufacturing-types/src/types/supplier.ts
import { z } from 'zod';

export const SupplierTypeSchema = z.enum([
  'cnc_shop',
  'pcb_fab',
  'sheet_metal_vendor',
  'injection_molding',
  'assembly_house',
  'logistics_provider',
  'material_supplier',
  'finishing_shop',
  'testing_lab',
  'additive_manufacturing',
  'robotics_integrator',
  'electronics_supplier',
]);

export type SupplierType = z.infer<typeof SupplierTypeSchema>;

export const SupplierCapabilitySchema = z.object({
  id: z.string(),
  name: z.string(),
  type: z.string(),
  process: z.string().optional(),
  min_order_quantity: z.number(),
  max_annual_capacity: z.number(),
  lead_time_days: z.object({
    standard: z.number(),
    expedited: z.number().optional(),
  }),
  cost_per_unit_base: z.number(),
  volume_discounts: z.array(z.object({
    min_quantity: z.number(),
    discount_percent: z.number(),
  })).optional(),
  precision_capability_microns: z.number().optional(),
  material_capabilities: z.array(z.string()).optional(),
  certifications: z.array(z.string()),
  properties: z.record(z.string(), z.unknown()).optional(),
});

export type SupplierCapability = z.infer<typeof SupplierCapabilitySchema>;

export const SupplierSchema = z.object({
  id: z.string(),
  name: z.string(),
  type: SupplierTypeSchema,
  location: z.object({
    country: z.string(),
    region: z.string(),
    city: z.string(),
  }),
  contact_email: z.string().email(),
  contact_phone: z.string().optional(),
  capabilities: z.array(SupplierCapabilitySchema),
  quality_score: z.number().min(0).max(100),
  reliability_score: z.number().min(0).max(100),
  cost_competitiveness_score: z.number().min(0).max(100),
  on_time_delivery_rate: z.number().min(0).max(100),
  defect_rate: z.number().min(0).max(100),
  minimum_order_value: z.number().optional(),
  payment_terms: z.string(),
  lead_time_variability: z.number(),
  certifications: z.array(z.string()),
  quality_audits: z.array(z.object({
    date: z.date(),
    score: z.number(),
    auditor: z.string(),
  })).optional(),
  properties: z.record(z.string(), z.unknown()).optional(),
  created_at: z.date(),
  updated_at: z.date(),
});

export type Supplier = z.infer<typeof SupplierSchema>;

export const SupplierQuoteSchema = z.object({
  id: z.string(),
  supplier_id: z.string(),
  part_id: z.string(),
  capability_id: z.string(),
  quantity: z.number(),
  unit_price: z.number(),
  total_price: z.number(),
  lead_time_days: z.number(),
  minimum_order_quantity: z.number(),
  volume_available: z.number(),
  expiration_date: z.date(),
  terms: z.string().optional(),
  created_at: z.date(),
  updated_at: z.date(),
});

export type SupplierQuote = z.infer<typeof SupplierQuoteSchema>;

export const ProcurementOrderSchema = z.object({
  id: z.string(),
  supplier_id: z.string(),
  supplier_quote_id: z.string(),
  status: z.enum(['draft', 'sent', 'confirmed', 'in_production', 'shipped', 'delivered', 'cancelled']),
  items: z.array(z.object({
    part_id: z.string(),
    quantity: z.number(),
    unit_price: z.number(),
  })),
  total_value: z.number(),
  expected_delivery_date: z.date(),
  actual_delivery_date: z.date().optional(),
  payment_status: z.enum(['pending', 'partial', 'paid', 'overdue']),
  created_at: z.date(),
  updated_at: z.date(),
});

export type ProcurementOrder = z.infer<typeof ProcurementOrderSchema>;

File: apps/api/app/schemas/supplier.py
"""Supplier schemas"""

from pydantic import BaseModel, Field, EmailStr
from uuid import UUID
from datetime import datetime
from decimal import Decimal
from typing import Optional, List


class SupplierCapabilityBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    type: Optional[str] = None
    process: Optional[str] = None
    min_order_quantity: Optional[int] = None
    max_annual_capacity: Optional[float] = None
    lead_time_standard_days: Optional[int] = None
    lead_time_expedited_days: Optional[int] = None
    cost_per_unit_base: Optional[Decimal] = Field(None, decimal_places=4)
    precision_capability_microns: Optional[float] = None
    material_capabilities: List[str] = []
    certifications: List[str] = []


class SupplierCapabilityCreate(SupplierCapabilityBase):
    pass


class SupplierCapabilityRead(SupplierCapabilityBase):
    id: UUID
    supplier_id: UUID

    class Config:
        from_attributes = True


class SupplierBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    type: str = Field(..., min_length=1, max_length=50)
    country: str = Field(..., min_length=1, max_length=50)
    region: Optional[str] = Field(None, max_length=100)
    city: Optional[str] = Field(None, max_length=100)
    contact_email: Optional[EmailStr] = None
    contact_phone: Optional[str] = Field(None, max_length=20)
    quality_score: int = Field(50, ge=0, le=100)
    reliability_score: int = Field(50, ge=0, le=100)
    cost_competitiveness_score: int = Field(50, ge=0, le=100)
    on_time_delivery_rate: float = Field(80.0, ge=0, le=100)
    defect_rate: float = Field(2.0, ge=0, le=100)
    minimum_order_value: Optional[Decimal] = Field(None, decimal_places=2)
    payment_terms: Optional[str] = None
    lead_time_variability: Optional[float] = None
    certifications: List[str] = []


class SupplierCreate(SupplierBase):
    capabilities: Optional[List[SupplierCapabilityCreate]] = None


class SupplierRead(SupplierBase):
    id: UUID
    created_at: datetime
    updated_at: datetime
    capabilities: List[SupplierCapabilityRead] = []

    class Config:
        from_attributes = True


class SupplierUpdate(BaseModel):
    name: Optional[str] = None
    type: Optional[str] = None
    country: Optional[str] = None
    region: Optional[str] = None
    city: Optional[str] = None
    contact_email: Optional[EmailStr] = None
    contact_phone: Optional[str] = None
    quality_score: Optional[int] = None
    reliability_score: Optional[int] = None
    cost_competitiveness_score: Optional[int] = None
    on_time_delivery_rate: Optional[float] = None
    defect_rate: Optional[float] = None
    minimum_order_value: Optional[Decimal] = None
    payment_terms: Optional[str] = None
    lead_time_variability: Optional[float] = None
    certifications: Optional[List[str]] = None


class SupplierQuoteBase(BaseModel):
    quantity: int = Field(..., gt=0)
    unit_price: Decimal = Field(..., decimal_places=4)
    total_price: Decimal = Field(..., decimal_places=2)
    lead_time_days: Optional[int] = None
    minimum_order_quantity: Optional[int] = None
    volume_available: Optional[float] = None
    expiration_date: Optional[datetime] = None
    terms: Optional[str] = Field(None, max_length=2000)


class SupplierQuoteCreate(SupplierQuoteBase):
    supplier_id: UUID
    part_id: UUID
    capability_id: Optional[UUID] = None


class SupplierQuoteRead(SupplierQuoteBase):
    id: UUID
    supplier_id: UUID
    part_id: UUID
    capability_id: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
